docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+  {
+    self.weight += food;
+  }
+
+  poop(self: Capibara)
+  {
+    self.weight -= 0.1; // Tiny poop!
+  }
+};
+```
+
+On the example above, the `newborn` method creates a `Capibara` with a specific name and a default weight.
+The `name` can't change, so we declare it with `let`, while the `weight` does change, so it's a `var`.
+
+There's nothing special about the method `newbord` in comparison with others, apart from the fact that it allocates, initialises and returns an object of the same type.
+
+Destructors are still under debate, but they won't have to deallocate memory like in C++, just execute finalization tasks like closing file handles, running some cleanup routines, checking final consistency, etc.
+
+## Primitive Types
+
+Verona treats numeric types specially in the compiler but their representaiton isn't special in the language.
+
+Verona's numeric types are singleton types (no fields) that have their machine representaiton based on their names.
+So, `I32` is a signed integer with 32-bit width (in LLVM's parlour, an `i32`) and so on.
+
+The full list of primitive numeric types are:
+ * Unsigned integers: `U8`, `U16`, `U32`, `U64` and `U128`
+ * Signed integers: `I8`, `I16`, `I32`, `I64` and `I128`
+ * Floating Point: `F32` and `F64`
+ * System types: `ULength` and `ILength` (machine-word sized integers)
+
+Note, integer literals have no defined type and their types (and representation) will be inferred by the compiler depending on the type of the variable being assigned to.
+
+## Type Unions
+
+Type unions means a variable can be either type on the list.
+
+For example:
+```ts
+foobar(a : (A | B | C))
+{
+  ...
+}
+
+main()
+{
+  let a : A;
+  let b : B;
+  let d : D;
+
+  // This is fine
+  foobar(a);
+  foobar(b);
+
+  // This is not
+  foobar(d); // ERROR
+}
+```
+
+The imlementation of `foobar` will have to cope with all three types, either via a `match` statement or taking advantage of some behaviour that is common to all three (say, an interface they all implement).
+
+The compiler is free, however, to optimise the calls by creating two versions of `foobar` and replace the calls to the generic `foobar` for their specific ones.
+If the semantics is defined at compile time, this can simplifies a `match` statement into their specific matches and even inline the call altogether.
+
+Example:
+```ts
+// The generic one, if there are still dynamic cases
+foobar(a : (A | B | C)) { ... }
+
+// The specific specialisations used in defined cases
+foobar(a : A) { ... }
+foobar(b : B) { ... }
+```
+
+## Type Intersections
+
+Intersection means a type must implement all guarantees of all types involved.
+
+For example:
+```ts
+interface Serialisable { toString() : String; };
+
+main()
+{
+  let x : (A & Serialisable & mut) = ...;
+}
+```
+
+In the above example, the type of `x` must conform to both `A` and `Serialisable` guarantees (methods, fields) as well as the `mutable` capability.
+
+Every type in Verona must have a capability, so:
```

Comment:
earlier you said that a capability is a type. here, with this phrasing you are saying that a capability is a thing that a type has.

Reply:
Hm, vero. I'm not sure how to phrase this, but every type must be the intersection of a type and a capability, but there shouldn't be two capabilities, but capability is a type, and intersections between types are allowed. It's really confusing to me.

Reply:
I think we're quite fuzzy about what we mean when we say 'capability'.  To me, the things that we're calling capabilities are actually permissions.  A capability is an unforgeable, delegable, token of authority that grants rights to perform some action.  In a capability system you must present present a capability with the requisite rights when you attempt to perform an action.  A Verona pointer *is* a capability.  You cannot create one from nothing.  Owning a pointer allows you to do things to the object that it refers to.  Specific permissions on the pointer confer the ability to do specific things (e.g. `mut` or `iso` permit mutation, `iso` or `imm` permit sending to a different behaviour, `imm` permits sharing with other behaviours).

Reply:
This is what capabilities meant to me, too. But in the type system, capabilities are also types that you can use with type inference, for example:
```
foo (a : imm) { ... }
```
Obviously, the type of `a` isn't just `imm` but this requires whatever type is passed to be viewed as having the `imm` capability. If `imm` wasn't a type, that statement would be invalid because `a` wouldn't have a type.

Also, if capabilities are types and we have type unions and intersections, then we can obviously use capabilities there, and then we naturally arrive at `Type & cap` that is required for all pointers.

I imagine capabilities can't be mixed in like other types (ex. `imm` & `mut`) not because they're special, but because their semantics (like other types can have too) don't mix. If you try to mix two types that have no intersection like `A & B`, then the resulting type is void.

Perhaps that's what it means to intersect capabilities?

But also, capabilities have intrinsic semantics that no other types have, so in a way, capabilities are special types. I don't know if this is a semantic difference (in the definition of the type system) or an implementation difference (like how numeric types are different than other singleton types).